pkg/ext/apiserver.go

```diff
@@ -0,0 +1,259 @@
+)
+
+var (
+	schemeBuilder = runtime.NewSchemeBuilder(addKnownTypes, metainternalversion.AddToScheme)
+	AddToScheme   = schemeBuilder.AddToScheme
+)
+
+func addKnownTypes(scheme *runtime.Scheme) error {
+	metav1.AddToGroupVersion(scheme, schema.GroupVersion{Version: "v1"})
+	return nil
+}
+
+type ExtensionAPIServerOptions struct {
+	// GetOpenAPIDefinitions is collection of all definitions. Required.
+	GetOpenAPIDefinitions             openapicommon.GetOpenAPIDefinitions
+	OpenAPIDefinitionNameReplacements map[string]string
+
+	// Authenticator will be used to authenticate requests coming to the
+	// extension API server. Required.
+	Authenticator authenticator.Request
+
+	// Authorizer will be used to authorize requests based on the user,
+	// operation and resources. Required.
+	//
+	// Use [NewAccessSetAuthorizer] for an authorizer that uses Steve's access set.
+	Authorizer authorizer.Authorizer
+
+	// Listener is the TCP listener that is used to listen to the extension API server
+	// that is reached by the main kube-apiserver. Required.
+	Listener net.Listener
+}
+
+// ExtensionAPIServer wraps a [genericapiserver.GenericAPIServer] to implement
+// a Kubernetes extension API server.
+//
+// Use [NewExtensionAPIServer] to create an ExtensionAPIServer.
+//
+// Use [InstallStore] to add a new resource store onto an existing ExtensionAPIServer.
+// Each resources will then be reachable via /apis/<group>/<version>/<resource> as
+// defined by the Kubernetes API.
+//
+// When Run() is called, a separate HTTPS server is started. This server is meant
+// for the main kube-apiserver to communicate with our extension API server. We
+// can expect the following requests from the main kube-apiserver:
+//
+// <path>                 <user>                 <groups>
+// /openapi/v2            system:aggregator      [system:authenticated]
+// /openapi/v3            system:aggregator      [system:authenticated]
+// /apis                  system:kube-aggregator [system:masters system:authenticated]
+// /apis/ext.cattle.io/v1 system:kube-aggregator [system:masters system:authenticated]
+type ExtensionAPIServer struct {
+	codecs serializer.CodecFactory
+	scheme *runtime.Scheme
+
+	genericAPIServer *genericapiserver.GenericAPIServer
+	apiGroups        map[string]genericapiserver.APIGroupInfo
+
+	authorizer authorizer.Authorizer
+
+	handlerMu sync.RWMutex
+	handler   http.Handler
+}
+
+type emptyAddresses struct{}
+
+func (e emptyAddresses) ServerAddressByClientCIDRs(clientIP net.IP) []metav1.ServerAddressByClientCIDR {
+	return nil
+}
+
+func NewExtensionAPIServer(scheme *runtime.Scheme, codecs serializer.CodecFactory, opts ExtensionAPIServerOptions) (*ExtensionAPIServer, error) {
+	if opts.Authenticator == nil {
+		return nil, fmt.Errorf("authenticator must be provided")
+	}
+
+	if opts.Authorizer == nil {
+		return nil, fmt.Errorf("authorizer must be provided")
+	}
+
+	if opts.Listener == nil {
+		return nil, fmt.Errorf("listener must be provided")
+	}
```

Comment:
Yup, as said here https://github.com/rancher/steve/pull/287/files#r1793733154 we need it even though we're not using it. Annoying, but not much we can do for now. It will become more useful in the next PR. Resolving :+1: